File: leetcode/binary-search/1658-Minimum-Operations-to-Reduce-X-to-Zero.java
class Solution {
    public int minOperations(int[] nums, int x) {
        int n = nums.length;

        // C++: accumulate(nums.begin(), nums.end(), 0)
        int total = 0;
        for (int num : nums) {
            total += num;
        }

        int k = total - x;

        if (k < 0)
            return -1;

        if (k == 0)
            return n;

        int maxi = 0;

        // C++: unordered_map<int, int>
        HashMap<Integer, Integer> mpp = new HashMap<>();

        // prefix sum 0 occurs before index 0
        mpp.put(0, -1);

        int prefix = 0;

        for (int i = 0; i < n; i++) {

            prefix += nums[i];

            // Check if prefix - k exists
            if (mpp.containsKey(prefix - k)) {
                maxi = Math.max(maxi, i - mpp.get(prefix - k));
            }

            // Store only the first occurrence
            if (!mpp.containsKey(prefix)) {
                mpp.put(prefix, i);
            }
        }

        return maxi == 0 ? -1 : n - maxi;
    }
}

File: leetcode/binary-search/1658-Minimum-Operations-to-Reduce-X-to-Zero.cpp
class Solution {
public:
    int minOperations(vector<int>& nums, int x) {
        int n=nums.size();
        int sm=0;
        for(int i=0;i<n;i++)
        sm+=nums[i];
        int target=sm-x;
        if(target<0)
        return -1;
        if(target==0)
        return n;
        int l=0,s=0,maxiLen=-1;
        for(int r=0;r<n;r++)
        {
            s+=nums[r];
            while(s>target)
            {
                s-=nums[l];
                l++;
            }
            if(s==target)
            maxiLen=max(maxiLen,r-l+1);
        }
        if(maxiLen==-1)
        return -1;
        return n-maxiLen;
    }
};